### DISCUSSION TITLE
OffScreen - NETCore PublishSingleFile as Background Service

### DISCUSSION THREAD

**Discussion by vinothtgithub** [cefsharp/CefSharp#4161] [General]
- CefSharp.OffScreen.NETCore version 103.0.90

- x64 architecture
	
- .NET6
    
- Win10 operating system

- OffScreen
    
- Steps to reproduce the problem:
    - Created .NET6 Background Service project and copied the example code from "https://github.com/cefsharp/CefSharp.MinimalExample/CefSharp.MinimalExample.OffScreen/Program.cs". Configured to run in x64. When running this project in console mode, sample code works as expected. But when deployed as windows service, same sample code is not working as expected and throwing error during CEF library code execution,
    - The problem is reproducible using the [MinimalExample](https://github.com/cefsharp/CefSharp.MinimalExample) and sample project is available in https://github.com/vinothtgithub/cefbgservice 
- Expected output is screen shot of google search home page with key input "CefSharp Was Here!". But following error during service start
- _Application: CEFSharpDemo.Net6.WokerService.exe
CoreCLR Version: 6.0.622.26707
.NET Version: 6.0.6
Description: The process was terminated due to an internal error in the .NET Runtime.at IP 00007FF61C079BD4 (00007FF61BE30000) with exit code c0000005._
- _Faulting application name: CEFSharpDemo.Net6.WokerService.exe, version: 1.0.0.0, time stamp: 0x6283da22
Faulting module name: CEFSharpDemo.Net6.WokerService.exe, version: 1.0.0.0, time stamp: 0x6283da22
Exception code: 0xc0000005
Fault offset: 0x0000000000249bd4
Faulting process id: 0x5190
Faulting application start time: 0x01d8956193c5323b
Faulting application path: C:\Users\xxxxxx\source\repos\CEFSharpDemo.Net6.WokerService\CEFSharpDemo.Net6.WokerService\bin\Release\net6.0\publish\CEFSharpDemo.Net6.WokerService.exe
Faulting module path: C:\Users\xxxxxx\source\repos\CEFSharpDemo.Net6.WokerService\CEFSharpDemo.Net6.WokerService\bin\Release\net6.0\publish\CEFSharpDemo.Net6.WokerService.exe
Report Id: 10d98fbb-247d-40e5-83e1-030dbfb8ea0f
Faulting package full name: 
Faulting package-relative application ID:_

**Comment by amaitland** [comment]
What is the actual error?

What settings did you publish with?

What settings did you install the service with?

**Reply by vinothtgithub** [reply]
**Actual error obtained from event log are following:**
![image](https://user-images.githubusercontent.com/35638648/178564428-e9785e1f-e4c2-4d87-96e1-717dedda4624.png)

![image](https://user-images.githubusercontent.com/35638648/178564809-742ef0f6-52cb-484e-a391-68730470202c.png)

![image](https://user-images.githubusercontent.com/35638648/178564994-0f0eeb05-5503-4187-9161-ad386170967e.png)

**Windows Error Report:**
_Version=1
EventType=APPCRASH
EventTime=133020435355332065
ReportType=2
Consent=1
UploadTime=133020435382592188
ReportStatus=268435456
ReportIdentifier=2c610589-f1be-4e99-8ca5-6ba9f0c077f4
IntegratorReportIdentifier=10d98fbb-247d-40e5-83e1-030dbfb8ea0f
Wow64Host=34404
NsAppName=CEFSharpDemo.Net6.WokerService.exe
OriginalFilename=CEFSharpDemo.Net6.WokerService.dll
AppSessionGuid=00005190-0000-0026-3b32-c5936195d801
TargetAppId=W:00061cc62061a91f00294b6261dcf760182b00000000!0000e58b39d4bd87d521dbd72a2a1ec39f4bdec28956!CEFSharpDemo.Net6.WokerService.exe
TargetAppVer=2022//05//17:17:23:46!0!CEFSharpDemo.Net6.WokerService.exe
BootId=4294967295
TargetAsId=4467
IsFatal=1
EtwNonCollectReason=1
Response.BucketId=4a908927c85b1ce25262031d975d787f
Response.BucketTable=4
Response.LegacyBucketId=1324624666028767359
Response.type=4
Sig[0].Name=Application Name
Sig[0].Value=CEFSharpDemo.Net6.WokerService.exe
Sig[1].Name=Application Version
Sig[1].Value=1.0.0.0
Sig[2].Name=Application Timestamp
Sig[2].Value=6283da22
Sig[3].Name=Fault Module Name
Sig[3].Value=CEFSharpDemo.Net6.WokerService.exe
Sig[4].Name=Fault Module Version
Sig[4].Value=1.0.0.0
Sig[5].Name=Fault Module Timestamp
Sig[5].Value=6283da22
Sig[6].Name=Exception Code
Sig[6].Value=c0000005
Sig[7].Name=Exception Offset
Sig[7].Value=0000000000249bd4
DynamicSig[1].Name=OS Version
DynamicSig[1].Value=10.0.19043.2.0.0.256.48
DynamicSig[2].Name=Locale ID
DynamicSig[2].Value=1033
DynamicSig[22].Name=Additional Information 1
DynamicSig[22].Value=5892
DynamicSig[23].Name=Additional Information 2
DynamicSig[23].Value=589267e8fd7f4ad761124d28f4676675
DynamicSig[24].Name=Additional Information 3
DynamicSig[24].Value=36ca
DynamicSig[25].Name=Additional Information 4
DynamicSig[25].Value=36ca1dd059a2bd7979892db6b1a05c6c
UI[2]=C:\Users\casee2009\source\repos\CEFSharpDemo.Net6.WokerService\CEFSharpDemo.Net6.WokerService\bin\Release\net6.0\publish\CEFSharpDemo.Net6.WokerService.exe
UI[5]=Close
UI[8]=CEFSharpDemo.Net6.WokerService stopped working and was closed
UI[9]=A problem caused the application to stop working correctly. Windows will notify you if a solution is available.
UI[10]=&Close
LoadedModule[0]=C:\Users\casee2009\source\repos\CEFSharpDemo.Net6.WokerService\CEFSharpDemo.Net6.WokerService\bin\Release\net6.0\publish\CEFSharpDemo.Net6.WokerService.exe
LoadedModule[1]=C:\WINDOWS\SYSTEM32\ntdll.dll
LoadedModule[2]=C:\Windows\system32\hmpalert.dll
LoadedModule[3]=C:\WINDOWS\System32\KERNEL32.dll
LoadedModule[4]=C:\WINDOWS\System32\KERNELBASE.dll
LoadedModule[5]=C:\WINDOWS\System32\ADVAPI32.dll
LoadedModule[6]=C:\WINDOWS\System32\msvcrt.dll
LoadedModule[7]=C:\WINDOWS\System32\sechost.dll
LoadedModule[8]=C:\WINDOWS\System32\RPCRT4.dll
LoadedModule[9]=C:\WINDOWS\System32\ole32.dll
LoadedModule[10]=C:\WINDOWS\System32\ucrtbase.dll
LoadedModule[11]=C:\WINDOWS\System32\combase.dll
LoadedModule[12]=C:\WINDOWS\System32\GDI32.dll
LoadedModule[13]=C:\WINDOWS\System32\win32u.dll
LoadedModule[14]=C:\WINDOWS\System32\gdi32full.dll
LoadedModule[15]=C:\WINDOWS\System32\msvcp_win.dll
LoadedModule[16]=C:\WINDOWS\System32\USER32.dll
LoadedModule[17]=C:\WINDOWS\System32\OLEAUT32.dll
LoadedModule[18]=C:\WINDOWS\System32\SHELL32.dll
LoadedModule[19]=C:\WINDOWS\SYSTEM32\VERSION.dll
LoadedModule[20]=C:\WINDOWS\system32\vmwsci.dll
LoadedModule[21]=C:\WINDOWS\System32\PSAPI.DLL
LoadedModule[22]=C:\WINDOWS\System32\bcryptPrimitives.dll
LoadedModule[23]=C:\WINDOWS\SYSTEM32\kernel.appcore.dll
LoadedModule[24]=C:\WINDOWS\SYSTEM32\icu.dll
LoadedModule[25]=C:\WINDOWS\System32\BCrypt.dll
State[0].Key=Transport.DoneStage1
State[0].Value=1
OsInfo[0].Key=vermaj
OsInfo[0].Value=10
OsInfo[1].Key=vermin
OsInfo[1].Value=0
OsInfo[2].Key=verbld
OsInfo[2].Value=19043
OsInfo[3].Key=ubr
OsInfo[3].Value=1766
OsInfo[4].Key=versp
OsInfo[4].Value=0
OsInfo[5].Key=arch
OsInfo[5].Value=9
OsInfo[6].Key=lcid
OsInfo[6].Value=1033
OsInfo[7].Key=geoid
OsInfo[7].Value=244
OsInfo[8].Key=sku
OsInfo[8].Value=48
OsInfo[9].Key=domain
OsInfo[9].Value=1
OsInfo[10].Key=prodsuite
OsInfo[10].Value=256
OsInfo[11].Key=ntprodtype
OsInfo[11].Value=1
OsInfo[12].Key=platid
OsInfo[12].Value=10
OsInfo[13].Key=sr
OsInfo[13].Value=0
OsInfo[14].Key=tmsi
OsInfo[14].Value=221697972
OsInfo[15].Key=osinsty
OsInfo[15].Value=3
OsInfo[16].Key=iever
OsInfo[16].Value=11.789.19041.0-11.0.1000
OsInfo[17].Key=portos
OsInfo[17].Value=0
OsInfo[18].Key=ram
OsInfo[18].Value=8192
OsInfo[19].Key=svolsz
OsInfo[19].Value=198
OsInfo[20].Key=wimbt
OsInfo[20].Value=0
OsInfo[21].Key=blddt
OsInfo[21].Value=191206
OsInfo[22].Key=bldtm
OsInfo[22].Value=1406
OsInfo[23].Key=bldbrch
OsInfo[23].Value=vb_release
OsInfo[24].Key=bldchk
OsInfo[24].Value=0
OsInfo[25].Key=wpvermaj
OsInfo[25].Value=0
OsInfo[26].Key=wpvermin
OsInfo[26].Value=0
OsInfo[27].Key=wpbuildmaj
OsInfo[27].Value=0
OsInfo[28].Key=wpbuildmin
OsInfo[28].Value=0
OsInfo[29].Key=osver
OsInfo[29].Value=10.0.19041.1766.amd64fre.vb_release.191206-1406
OsInfo[30].Key=buildflightid
OsInfo[30].Value=CEDF9686-837B-4D8D-8779-782C4948D929.1
OsInfo[31].Key=edition
OsInfo[31].Value=Professional
OsInfo[32].Key=ring
OsInfo[32].Value=Retail
OsInfo[33].Key=expid
OsInfo[34].Key=fconid
OsInfo[35].Key=containerid
OsInfo[36].Key=containertype
OsInfo[37].Key=edu
OsInfo[37].Value=0
FriendlyEventName=Stopped working
ConsentKey=APPCRASH
AppName=CEFSharpDemo.Net6.WokerService
AppPath=C:\Users\casee2009\source\repos\CEFSharpDemo.Net6.WokerService\CEFSharpDemo.Net6.WokerService\bin\Release\net6.0\publish\CEFSharpDemo.Net6.WokerService.exe
NsPartner=windows
NsGroup=windows8
ApplicationIdentity=65A0070C1D27C61861773DB9FD1A0F40
MetadataHash=1028221732_

**Reply by vinothtgithub** [reply]
**Publish Settings:**
```
<?xml version="1.0" encoding="utf-8"?>
<!--
https://go.microsoft.com/fwlink/?LinkID=208121.
-->
<Project>
  <PropertyGroup>
    <Configuration>Release</Configuration>
    <Platform>x64</Platform>
    <PublishDir>bin\Release\net6.0\publish\</PublishDir>
    <PublishProtocol>FileSystem</PublishProtocol>
    <TargetFramework>net6.0</TargetFramework>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <SelfContained>true</SelfContained>
    <PublishSingleFile>true</PublishSingleFile>
    <PublishReadyToRun>true</PublishReadyToRun>
    <PublishTrimmed>false</PublishTrimmed>
  </PropertyGroup>
</Project>
```

**Reply by vinothtgithub** [reply]
Service installation has done using microsoft documentation https://docs.microsoft.com/en-us/dotnet/core/extensions/windows-service. 

For service Installation, open Windows Powershell in Administrator mode 

**[USE SCRIPT TO CREATE SERVICE]**
_sc.exe create "Cef Browser Background Service" binPath="C:\Users\casee2009\source\repos\CEFSharpDemo.Net6.WokerService\CEFSharpDemo.Net6.WokerService\bin\Release\net6.0\publish\CEFSharpDemo.Net6.WokerService.exe"_
**[OUTPUT]**
_CreateService SUCCESS_

**[USE SCRIPT TO START SERVICE]** **ERROR IS REPRODUCIBLE IN THIS STEP**
_sc.exe start "Cef Browser Background Service"_
**[OUTPUT]**
_SERVICE_NAME: Cef Browser Background Service
        TYPE               : 10  WIN32_OWN_PROCESS
        STATE              : 2  START_PENDING
                                (NOT_STOPPABLE, NOT_PAUSABLE, IGNORES_SHUTDOWN)
        WIN32_EXIT_CODE    : 0  (0x0)
        SERVICE_EXIT_CODE  : 0  (0x0)
        CHECKPOINT         : 0x0
        WAIT_HINT          : 0x7d0
        PID                : 21508
        FLAGS              :_

**Reply by amaitland** [reply]
I need the actual exception that's being thrown, not just the generic event log, attach the debugger to the process.

> Publish Settings:

You must self host the browser sub process to publish a single file exe. There are details on the publish settings required and self hosting at https://github.com/cefsharp/CefSharp/issues/3407

You'll need to change your example. I'd suggest you start without publishing a self contained single file build and get that working first.

> Service installation has done using microsoft documentation https://docs.microsoft.com/en-us/dotnet/core/extensions/windows-service.

The OffScreen example code is not meant to be run in a service and would require changes to the paths.

**Reply by vinothtgithub** [reply]
**Thanks. Like you advised I copied the files from "x64/Debug" folder (which is a working version) and installed as service. its now running as background service and working as expected.**

**Actual Exception captured from console screen:**

> _info: CEFSharpDemo.Net6.WokerService.Worker[0]
>       Starting service
> Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
> Repeat 2 times:
>    at <Module>._getFiberPtrId()
>    at <Module>.<CrtImplementationDetails>.LanguageSupport._Initialize(<CrtImplementationDetails>.LanguageSupport*)
>    at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize(<CrtImplementationDetails>.LanguageSupport*)
>    at <Module>..cctor()
>    at CefSharp.CefSettingsBase..ctor()
>    at CefSharp.OffScreen.CefSettings..ctor()
>    at CEFSharpDemo.Net6.WokerService.Worker+<<ExecuteAsync>b__2_0>d.MoveNext()
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[CEFSharpDemo.Net6.WokerService.Worker+<<ExecuteAsync>b__2_0>d, CEFSharpDemo.Net6.WokerService, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]](<<ExecuteAsync>b__2_0>d ByRef)
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[CEFSharpDemo.Net6.WokerService.Worker+<<ExecuteAsync>b__2_0>d, CEFSharpDemo.Net6.WokerService, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]](<<ExecuteAsync>b__2_0>d ByRef)
>    at CEFSharpDemo.Net6.WokerService.Worker.<ExecuteAsync>b__2_0()
>    at CEFSharpDemo.Net6.WokerService.AsyncContext.Run(System.Func`1<System.Threading.Tasks.Task>)
>    at CEFSharpDemo.Net6.WokerService.Worker.ExecuteAsync(System.Threading.CancellationToken)
>    at Microsoft.Extensions.Hosting.BackgroundService.StartAsync(System.Threading.CancellationToken)
>    at Microsoft.Extensions.Hosting.Internal.Host+<StartAsync>d__12.MoveNext()
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Extensions.Hosting.Internal.Host+<StartAsync>d__12, Microsoft.Extensions.Hosting, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]](<StartAsync>d__12 ByRef)
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.Extensions.Hosting.Internal.Host+<StartAsync>d__12, Microsoft.Extensions.Hosting, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]](<StartAsync>d__12 ByRef)
>    at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(System.Threading.CancellationToken)
>    at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions+<RunAsync>d__4.MoveNext()
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions+<RunAsync>d__4, Microsoft.Extensions.Hosting.Abstractions, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]](<RunAsync>d__4 ByRef)
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions+<RunAsync>d__4, Microsoft.Extensions.Hosting.Abstractions, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]](<RunAsync>d__4 ByRef)
>    at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(Microsoft.Extensions.Hosting.IHost, System.Threading.CancellationToken)
>    at Program+<<Main>$>d__0.MoveNext()
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Program+<<Main>$>d__0, CEFSharpDemo.Net6.WokerService, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]](<<Main>$>d__0 ByRef)
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Program+<<Main>$>d__0, CEFSharpDemo.Net6.WokerService, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]](<<Main>$>d__0 ByRef)
>    at Program.<Main>$(System.String[])
>    at Program.<Main>(System.String[])_ 

**I will refer the #3407 and make working version, then I will deploy it as background service**

> You must self host the browser sub process to publish a single file exe. There are details on the publish settings required and self hosting at https://github.com/cefsharp/CefSharp/issues/3407You must self host the browser sub process to publish a single file exe. There are details on the publish settings required and self hosting at https://github.com/cefsharp/CefSharp/issues/3407"

**Reply by amaitland** [reply]
Personally i see little benefit in publishing a single file for a service. Zipping the release output into a zip file which includes the build version in it's name and having a batch file for automation would be my preference. Much easier to troubleshoot.

https://docs.microsoft.com/en-us/visualstudio/msbuild/zipdirectory-task?view=vs-2022
